### ISSUE TITLE
Install gcp Pub/Sub plugin

### ISSUE THREAD

**Issue by bbinet** [mozilla-services/lua_sandbox_extensions#322]
I want to install GCP Pub/Sub plugin on debian stretch: are there some installation instructions somewhere?

I had a look to the [Dockerfile](https://github.com/mozilla-services/lua_sandbox_extensions/blob/87002d976a86170a02dc39e6a287320e04015dd4/Dockerfile#L31-L36), but it seems to be Redhat only: everything seem to be available from the grpc_stow.tgz archive, but how do you create this archive, and would it work for debian based systems?

**Comment by trink** [comment]
That tarball is all the Google tooling; you will have to build and install GRPC and the googleapis. 
https://github.com/grpc/grpc/blob/master/INSTALL.md
https://github.com/googleapis/googleapis

1) There will be some conflicts between the two installs that have to be manually resolved 
2) You should remove the shared libraries to force it to statically link as their pkg-config output will not produce what is needed in the GRPC_STATIC_LIBRARIES variable.

I have successfully built/tested on Ubuntu https://hsadmin.trink.com/packages/ubuntu16.04/luasandbox-gcp_0.0.1-1_amd64.deb

**Comment by bbinet** [comment]
Ok, I think this is because I'm using an old version of protobuf v3.0.0 backported from debian buster:
```
root@hl-wrk-1-dev:/tmp# dpkg -l | grep proto(shared library)
ii  libprotobuf-dev:amd64      3.0.0-9.1+hl~1                 amd64        protocol buffers C++ library (development files)
ii  libprotobuf-lite10:amd64   3.0.0-9.1+hl~1                 amd64        protocol buffers C++ library (lite version)
ii  libprotobuf10:amd64        3.0.0-9.1+hl~1                 amd64        protocol buffers C++ library
ii  libprotoc-dev:amd64        3.0.0-9.1+hl~1                 amd64        protocol buffers compiler library (development files)
ii  libprotoc10:amd64          3.0.0-9.1+hl~1                 amd64        protocol buffers compiler library
ii  protobuf-compiler          3.0.0-9.1+hl~1                 amd64        compiler for protocol buffer definition files
ii  protobuf-compiler-grpc     1.3.2-1+hl~1                   amd64        high performance general RPC framework - protobuf plugin
```
So I will try to build the latest 3.5.1 version instead.

Just another question: when luasandbox-gcp_0.0.1-1_amd64.deb has been generated, do we need to other dependencies (like libgrpc3) or everything needed is self contained in luasandbox-gcp_0.0.1-1_amd64.deb ?

**Comment by trink** [comment]
When statically linked it is self contained
```
ldd /usr/lib/luasandbox/io_modules/gcp.so 
	linux-vdso.so.1 =>  (0x00007ffe42d87000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f46a847f000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f46a8265000)
	libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007f46a7ffc000)
	libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007f46a7bb8000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f46a7836000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f46a752d000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f46a7317000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f46a6f4d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f46a8cb2000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f46a6d49000)
```

**Comment by bbinet** [comment]
Ok, thanks.
I've been able to build gcp when using shared libraries, but when I try to build statically (by removing shared libs as you advised. => `/usr/local/lib/*.so*`), I have the following error:
```
Scanning dependencies of target gcp
[ 32%] Building CXX object gcp/CMakeFiles/gcp.dir/pubsub.cpp.o
[ 33%] Building CXX object gcp/CMakeFiles/gcp.dir/usr/local/include/google/api/annotations.pb.cc.o
[ 33%] Building CXX object gcp/CMakeFiles/gcp.dir/usr/local/include/google/api/http.pb.cc.o
[ 34%] Building CXX object gcp/CMakeFiles/gcp.dir/usr/local/include/google/pubsub/v1/pubsub.pb.cc.o
[ 34%] Building CXX object gcp/CMakeFiles/gcp.dir/usr/local/include/google/pubsub/v1/pubsub.grpc.pb.cc.o
[ 35%] Linking CXX shared library gcp.so
/usr/bin/ld: //usr/local/lib/libprotobuf.a(arena.o): relocation R_X86_64_TPOFF32 against symbol `_ZN6google8protobuf8internal9ArenaImpl13thread_cache_E' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: //usr/local/lib/libprotobuf.a(descriptor.o): relocation R_X86_64_PC32 against symbol `_ZTVN6google8protobuf8internal16FunctionClosure0E' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
gcp/CMakeFiles/gcp.dir/build.make:199: recipe for target 'gcp/gcp.so' failed
make[2]: *** [gcp/gcp.so] Error 1
CMakeFiles/Makefile2:1859: recipe for target 'gcp/CMakeFiles/gcp.dir/all' failed
make[1]: *** [gcp/CMakeFiles/gcp.dir/all] Error 2
make: *** [all] Error 2
```
The error says I should recompile with  `-fPIC` option, so I tried to add `set(CMAKE_POSITION_INDEPENDENT_CODE ON)` line to `lua_sandbox_extensions/gcp/CMakeLists.txt`, but same error.
Any idea?

**Comment by bbinet** [comment]
I'm currently trying to rebuild protobuf with `-fPIC` option to see if it can fix my issue

**Comment by bbinet** [comment]
That was it.
I've now successfully built gcp extension statically after compiling protobuf with `-fPIC`.

Thanks for you support!

**Comment by bbinet** [comment]
But I've noticed that my `gcp.so` shared lib has an additional dependency on `libc-ares2` that you don't have:
```
$ ldd usr/lib/luasandbox/io_modules/gcp.so
	linux-vdso.so.1 (0x00007fff3bba0000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0666cca000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0666ab0000)
	libcares.so.2 => /usr/lib/x86_64-linux-gnu/libcares.so.2 (0x00007f066689e000)
	libssl.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.2 (0x00007f0666635000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0666431000)
	libcrypto.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2 (0x00007f0665fcb000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0665c49000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0665945000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f066572e000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f066538f000)
	/lib64/ld-linux-x86-64.so.2 (0x000055bb57db8000)
```
Is it normal?

**Comment by trink** [comment]
I wouldn't be surprised if the RPC code pulls it in (it is just not available on my 16.04 box). I am glad you are good to go now.
